### PR TITLE
Add fp16 test to `test_n_step_rnn`

### PR DESCRIPTION
**Pull request by belldandyxtq**
This commit adds the fp16 tests to the n_step_rnn tests.
The chainer.testing is used to parameterize and switch between the
float16 and float32.

The old `chainer.cuda` is replaced with `chainer.backends.cuda`

### CODE REVIEW

**Comment by crcrpar** [comment]
I think this is a duplicate of #5808 to some extent.

**Comment by belldandyxtq** [comment]
@crcrpar This PR adds the tests for the `n_step_rnn` in `ChainerMN` tests, which is not covered by #5808. But it seems  this PR depends on #5808 to support fp16 and no changes need to be made on the n_step_rnn in `ChainerMN`

**Comment by shu65** [comment]
Jenkins, test this please.

File: tests/chainermn_tests/links_tests/test_n_step_rnn.py [hunk 1]
```diff
@@ -39,93 +39,109 @@ def __call__(self, xs, ts):
...
 
-    @chainer.testing.attr.gpu
-    @pytest.mark.filterwarnings('ignore::DeprecationWarning')
-    def test_homogeneous_rnn_gpu(self):
-        self.check_homogeneous_rnn(True)
 
-    def check_heterogeneous_rnn(self, gpu):
-        self.setup(gpu)
+@chainer.testing.attr.gpu
+@pytest.mark.filterwarnings('ignore::DeprecationWarning')
```
Comment (shu65) [inline tests/chainermn_tests/links_tests/test_n_step_rnn.py]: Why this filter is needed?

Comment (belldandyxtq) [reply tests/chainermn_tests/links_tests/test_n_step_rnn.py]: These are legacy filters from https://github.com/chainer/chainer/commit/bbbcf9358ea1b5cc156a360853eb0bbd7beb89bf
to surpass the deprecated `add_link` https://github.com/chainer/chainer/blob/76a4aeef7ef86be8c273995b34338f89d00273f7/chainer/link.py#L920
The deprecated mark is removed, so I also remove the filter. There is another file affected 
https://github.com/chainer/chainer/blob/307a858b053e86b140fa71432fa60950dc83bee6/tests/chainermn_tests/links_tests/test_multi_node_chain_list.py#L314
I will open an issue for it

**Comment by shu65** [comment]
Jenkins, test this please.

**Pull request merged**